### ISSUE TITLE
Add input that accepts all input data formats using HTTP

### ISSUE THREAD

**Issue by danielnelson** [influxdata/telegraf#3226]
## Feature Request

The current `http_listener` input only accepts influx data format and attempts to behave like an InfluxDB server with respect to parsing, paths, headers, etc.

### Proposal:

Add a plugin alias for the current http_listener plugin: `influxdb_listener`, at some point in the future we remove the old http_listener.

Create a new input from scratch called `http_listener_ng`, or a good name if anyone has ideas.  The initial configuration could look something like:

```toml
[[inputs.http_listener_ng]]
  address = ":8186"

  paths = ["/telegraf"]
  methods = ["POST", "PUT"]

  read_timeout = "10s"
  write_timeout = "10s"

  tls_cert= "/etc/telegraf/server/cert.pem"
  tls_key = "/etc/telegraf/server/key.pem"
  tls_allowed_cacerts = ["/etc/telegraf/clientca.pem"]

  data_format = "influx"
```

### Current behavior:

No plugin exists that accepts the input data formats over HTTP.

### Desired behavior:

Plugin exists.

### Use case: [Why is this important (helps with prioritizing requests)]

Allows pushing application metrics to Telegraf over http in any format.

This is commonly needed:
- https://github.com/influxdata/telegraf/pull/3216
- https://community.influxdata.com/t/json-http-plugin-es/2313

**Comment by jul1u5** [comment]
It has been over a year now. Are there any updates regarding this feature?

**Comment by jul1u5** [comment]
I would be willing to implement this feature myself. I assume I could just copy the source of the current `http_listener` and then implement the proposed config options.

**Comment by glinton** [comment]
I don't think we recommend copying the http_listener plugin, but if you'd like to make this a reality that would be fantastic! If you need anything from us, let me know.

**Comment by jul1u5** [comment]
I am currently trying to run/test my plugin. I executed `make static`, but when I tried to run `make plugin-http_plugin_ng` I got an error that there is no such file `dev/docker-compose.yml`. I am not sure how to set up docker-compose with my plugin on the telegraf image.

**Comment by glinton** [comment]
To build, I just run `make telegraf` or `go build ./cmd/telegraf`

**Comment by jul1u5** [comment]
I have a couple questions about the implementation:
How should the plugin deal with `paths` config option? Should it add a tag containing the request's path to the newly created metric?
And what about `methods` option? Right now I just check if the request's method matches the ones specified in the config and return bad request if it doesn't match.

**Comment by danielnelson** [comment]
In retrospect, I think we should just have a single path, I think most of the time you will just need a single path and there is always the option to define the plugin multiple times.  Let's not add any new tags by default, I think for the plugins that use `data_format` we should avoid modifying the data by default.
```
path = "/telegraf"
```

For `methods`, if the request method doesn't match then return `405 Method Not Allowed`.

To add a bit more context about what we what we would like changed from the current `http_listener`, the main issue with it is the bufferpool and how it chunks the incoming data.  Both of these features we would like to remove unless it turns out they are critical to performance.

**Comment by jul1u5** [comment]
So looks like I have finished implementing the plugin. I tested with `http_plugin` tests and added one with `json`. They all pass.

I have minimally removed the buffering: `bufferpool.go` is removed, but buffering process is still mostly the same. I tried to experiment by simplifying it but some tests failed. I could try more if this is unsatisfactory.

By the way, I have also removed `precision` option (which is in `http_plugin`) for pushed metrics as I think it's not needed for this plugin now that it accepts multiple data formats. 

If everything is okay, I can create the pull request. To which branch should I merge?
